lib/Conversion/TorchToLinalg/Pooling.cpp

```diff
@@ -629,159 +629,177 @@ This is problematic for linalg ops for a few reasons:
 h! Although it is a bit like using a hammer to paint, our workaround is to use
 tensor.extract to access the elements of the input tensor inside our linalg
 generic op's payload.
-
-Current TODO's:
-    1. gather most of the boilerplate out of this op and make it into an
-adaptive pooling helper function.
-    2. figure out what to do with the conflicting decompositions in
-DecomposeComplexOps.cpp
-    3. Implement more efficient passes for when the kernel-size, input spatial
-dims, and output spatial dims are constant.
 */
 
 namespace {
-class ConvertAtenAdaptiveAvgPool1dOp
-    : public OpConversionPattern<AtenAdaptiveAvgPool1dOp> {
+
+class AdaptivePoolingHelper {
 public:
-  using OpConversionPattern::OpConversionPattern;
-  LogicalResult
-  matchAndRewrite(AtenAdaptiveAvgPool1dOp op, OpAdaptor adaptor,
-                  ConversionPatternRewriter &rewriter) const override {
+  AdaptivePoolingHelper(ConversionPatternRewriter &cpr, int64_t rnk,
+                        int64_t nsp, Type elt)
+      : rewriter(cpr), rank(rnk), nonSpatial(nsp), elementType(elt) {}
+
+  // Variables that are used in various helper functions in the derived classes
+  // are stored as members of the base class (to reduce the number of arguments
+  // passed to helper functions).
+  ConversionPatternRewriter &rewriter;
+  const int64_t rank;
+  const int64_t nonSpatial;
+  Type elementType;
+};
 
-    Location loc = op->getLoc();
-    const TypeConverter *typeConverter = getTypeConverter();
+template <class OpTy>
+class AdaptiveMaxPoolingHelper : public AdaptivePoolingHelper {
 
-    // get rank of input (same as rank of output)
-    int64_t rank =
-        adaptor.getSelf().getType().cast<RankedTensorType>().getRank();
-    // input operand should be NCH (i.e. rank 3)
-    if (rank != 3) {
-      return rewriter.notifyMatchFailure(op, "only supports input type NCH");
+  // This member variable is templated, so I've chosen not to make it part of
+  // the base class (to keep the base class non-templated).
+  const OpConversionPattern<OpTy> &opConversionPattern;
+
+public:
+  // Constructor for AdaptiveMaxPoolingHelper. Just forwards all arguments
+  // (except the OpConversionPattern) to the base class constructor.
+  template <typename... Args>
+  AdaptiveMaxPoolingHelper(const OpConversionPattern<OpTy> &ocp, Args &&...args)
+      : AdaptivePoolingHelper(std::forward<Args>(args)...),
+        opConversionPattern(ocp) {}
+
+  LogicalResult auxTensorSetup(OpTy op, const SmallVector<Value> &outputSizes,
```

Comment:
Hi, the auxiliary tensor is a tensor constructed in the first linalg generic op that is not the pooling operation. Both adaptive avg and adaptive max need some auxiliary tensor. I'm doing avg pooling first with a sum pool so as not to repeatedly compute and divide by the kernel volume on each iteration of the loop; therefore, I use the aux tensor keep track of what the kernel volume is for the sum pooling operation (so that it can later tensor divide sumpool/kernelVol). For max pooling, the function unfortunately always needs to return indices, so the indices (i.e., the integer position in the flattened input at which the local max occurs) need to be recorded in an aux tensor.

I offloaded this logic to a helper function to facilitate switching between those two uses (or possibly not doing anything if no aux tensor is required for a future op).